fcrepo-exts/fcrepo-migration-validator
# Validation for deleted datastreams

**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3686

# What does this Pull Request do?

Handle datastreams which have been marked as deleted and check that OCFL has a deleted version.

# What's new?

* Validation for deleted object resource if a F3 object resource is marked as deleted
* When an object resource is deleted, remove it from the HEAD datastreams
* Track number of deleted versions to re-align F3 and F6 version numbers
* Add enum for object resource deleted
* Add fixture data for tests to cover deleted object resources
* Add cli option for treating `Inactive` states as deleted
* Update PropretyResolver to return an Optional in the event a property cannot be found
* Update PropertyResolver to provide a fallback method for getting properties through n-triples

# How should this be tested?

Run a migration on a dataset with deleted object resources then run the validator on the migrated data. When checking output, there should be entries with `SOURCE_OBJECT_RESOURCE_DELETED` entries.

# Additional Notes:

* I shuffled around some of the logic for setting up validations to handle all Delete operations at the same time
* I _think_ `Deleted` versions will always be the terminal version of an Object Resource, but I'm not 100% on it. When trying to create an Object Resource which did alternate between Active -> Deleted -> Active, I noticed that the Deleted version was ignored (though tracked in the Audit Object resource). If this is the case, some of the code can be updated to be a bit simpler.
* The enum type could probably be named better, maybe just `OBJECT_RESOURCE_DELETED` rather than trying to differentiate between source/target

# Interested parties

@dbernstein
